source/common/conn_pool/conn_pool_base.cc

```diff
@@ -208,22 +208,23 @@ void ConnPoolImplBase::onStreamClosed(Envoy::ConnectionPool::ActiveClient& clien
                                       bool delay_attaching_stream) {
   ENVOY_CONN_LOG(debug, "destroying stream: {} remaining", client, client.numActiveStreams());
   ASSERT(num_active_streams_ > 0);
-  // Reflect there's one less stream in flight.
-  bool had_negative_capacity = client.hadNegativeDeltaOnStreamClosed();
   state_.decrActiveStreams(1);
   num_active_streams_--;
   host_->stats().rq_active_.dec();
   host_->cluster().stats().upstream_rq_active_.dec();
   host_->cluster().resourceManager(priority_).requests().dec();
-  // If the effective client capacity was limited by concurrency, increase connecting capacity.
-  // If the effective client capacity was limited by max total streams, this will not result in an
-  // increment as no capacity is freed up.
-  // We don't update the capacity for HTTP/3 as the stream count should only
-  // increase when a MAX_STREAMS frame is received.
-  if (trackStreamCapacity() && (client.remaining_streams_ > client.concurrent_stream_limit_ -
-                                                                client.numActiveStreams() - 1 ||
-                                had_negative_capacity)) {
-    state_.incrConnectingAndConnectedStreamCapacity(1);
+  // We don't update the capacity for HTTP/3 as the stream count should only.
+  if (trackStreamCapacity()) {
+    // If the effective client capacity was limited by concurrency, increase connecting capacity.
+    bool limited_by_concurrency =
+        client.remaining_streams_ > client.concurrent_stream_limit_ - client.numActiveStreams() - 1;
+    // The capacity calculated by concurrency could be negative (maybe explicitly call out SETTINGS
```

Comment:
"maybe explicitly call out SETTINGS frame"
Oops, I didn't mean this one to be a literal suggestion :-P

// The capacity calculated by concurrency could be negative if a SETTINGS frame lowered the number of
// allowed streams. In this case, effective client capacity was still limited by concurrency, compare
// client.concurrent_stream_limit_ and client.numActiveStreams() directly to avoid overflow.